### PR TITLE
OKFN proposal

### PR DESCRIPTION
**Pull request by lwm**
I'll add links to relevant resources when I paste it into the google doc style submission form.

### CODE REVIEW

**Comment by lwm** [comment]
@tomchristie 

A number of people have reviewed this with me. 

I'd still like to get your input before moving forward.

**Comment by tomchristie** [comment]
Looking over it now! 😄

File: posts/okfn.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+This is potentially a **huge** audience.
+
+By providing generated documentation and client libraries bindings the creator
+of a web API will be able to better focus on creating a high quality API. The
+workload in providing these expected features is greatly reduced.
+Documentation and client library maintenance effort is also greatly reduced.
+
+# Have you ever been working on the idea? If so, briefly describe the current state and explain the new feature.
+
+I have not been working on this idea. I have been exploring this proposal with
```
Comment (tomchristie) [inline posts/okfn.md]: Could drop "I have not been working on this idea." - You're already undertaking design work on this.

File: posts/okfn.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+it a worthwhile project to undertake.
+
+ReDoc is a project which is providing nicely formatted documentation for a
+single web API schema format. It is being developed with a permissive license
+which could be used as a good starting point for this project.
+
+# How many hours will you probably need during a period of 6 months for the implementation of the project idea?
+
+I would estimate to work full time (~35 hours per week) on this project for the
+duration of the 6 months.
```
Comment (tomchristie) [inline posts/okfn.md]: Yup, seems reasonable. You could consider breaking it down (eg 4 client libs at 1 month each, plus the documentation tool itself at 2 months) But I don't think you *necessarily need to*.  If you do break it down, then keep it very brief.

File: posts/okfn.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+# Sketch briefly the most important milestones of your project.
+
+  - Develop new mainstream programming language client libraries
+    - Java, Ruby, Php, Node, Go
```
Comment (tomchristie) [inline posts/okfn.md]: Could under-promise here, eg. pick the 3 you'd most prefer to work on.
Would be easy to take on too much.

File: posts/okfn.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+coreapi-doc, a web API documentation tool and client libraries.
+
+# What problem do you want to solve with your project? What was your motivation?
+
+Providing documentation and client libraries for web APIs is an extremely
+difficult task but critical to wide adoption and consumer satisfaction. It has
+been until now only available to groups with significant time and manpower to
+create such solutions. Even when those solutions are created, they are ad hoc
+and dependent on that single schema/hypermedia format. Those solutions are not
```
Comment (tomchristie) [inline posts/okfn.md]: I would not say "dependent on that single schema/hypermedia format" - most API documentation and client tooling isn't schema-driven, so in fact they're only usable with a single API.

File: posts/okfn.md [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+extensible.
+
+There are two problems here:
+
+  - Web APIs are exposed in many different representations. Client libraries
+    must be written exclusively for those specific representations. These
```
Comment (tomchristie) [inline posts/okfn.md]: Again, I'd change this around. Client libraries are typically written against a single API (not against a schema representation) which is why it's such repetitive and costly work.

File: posts/okfn.md [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    documentation gives no guarantees that it is an exhaustive reference.
+    Documentation can easily go stale with regard to current development.
+
+I want to provide a tool that automatically generates documentation for any web
+API that exposes a schema/hypermedia format and provide ready made client
+libraries in various mainstream programming languages that will work out of the
+box.
+
+My motivation is coming from the developments of the djangorestframework and
+coreapi libraries. Coreapi gives developers a single representation for any web
```
Comment (tomchristie) [inline posts/okfn.md]: "Core API", not "Coreapi"

File: posts/okfn.md [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    libraries must be maintained, must be provided in many different
+    programming languages and require a synchronising effort to provide similar
+    functionality.
+
+  - A web API consumer can never be guaranteed that the documentation is up to
+    date unless the documentation is generated from the underlying web API. The
+    documentation gives no guarantees that it is an exhaustive reference.
+    Documentation can easily go stale with regard to current development.
+
+I want to provide a tool that automatically generates documentation for any web
```
Comment (tomchristie) [inline posts/okfn.md]: Perhaps "An API tool that includes both schema-driven documentation and schema-driven client libraries"

File: posts/okfn.md [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+My motivation is coming from the developments of the djangorestframework and
+coreapi libraries. Coreapi gives developers a single representation for any web
+API that exposes a schema/hypermedia format with which my tool can be driven
+by.
+
+# How will your project solve this problem?
+
+My project will rely on a coreapi Document as an internal representation. Any
+web API that exposes a schema/hypermedia format can be translated into a
```
Comment (tomchristie) [inline posts/okfn.md]: I'd suggest dropping any mention of hypermedia - confuses the message slightly. Let's focus specifically on schema-driven.

File: posts/okfn.md [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+box.
+
+My motivation is coming from the developments of the djangorestframework and
+coreapi libraries. Coreapi gives developers a single representation for any web
+API that exposes a schema/hypermedia format with which my tool can be driven
+by.
+
+# How will your project solve this problem?
+
+My project will rely on a coreapi Document as an internal representation. Any
```
Comment (tomchristie) [inline posts/okfn.md]: Perhaps start of without getting into the implementation 

1. First explain what you're doing *without* reference to Core API. "A schema-driven API documentation tool and client libraries"
2. Then explain that the work will be based on Core API, and how that'll allow the documentation tool to work with many different schema formats.

File: posts/okfn.md [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+
+# Sketch briefly the most important milestones of your project.
+
+  - Develop new mainstream programming language client libraries
+    - Java, Ruby, Php, Node, Go
+  - Develop a nicely formatted rendering of a coreapi Document
+  - Develop client library bindings for rendering
+  - Merge the documentation and library bindings into rendering
+  - Develop platform for hosted documentation service
+  - General community management (mailing lists, bug tracking, project management)
```
Comment (tomchristie) [inline posts/okfn.md]: Perhaps consider building this up in a slightly different direction. Eg. the client libraries should probably be one of the later things to tackle, as once you've got everything else in place you can then start picking them off one-by-one.

- A renderer class for Django REST framework that converts a `coreapi.Document` schema representation into an HTML document page.
- Add `curl` client documentation to the renderer.
- Add client documentation for the existing python client to the renderer.
- Release a hosted service, allowing users to enter a schema URL and view the resulting documentation.
- Develop schema-driven client libraries in Java, Swift & Go (*) using the Python implementation as a reference point.
- Add client library documentation for each client library to the project.
- General community management (mailing lists, bug tracking, project management)

(*) Or whatever you choose.

Comment (tomchristie) [inline posts/okfn.md]: In fact you could also finish off here by noting that "The various client libraries will be developed once the core project is in place, reducing the risk of delivering the project, and allowing more or fewer languages to be tackled as time allows."

**Comment by tomchristie** [comment]
Okay, hope that helps!

**Comment by lwm** [comment]
> Okay, hope that helps!

Awesome, thank you very much!!!

File: posts/okfn.md [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+
+Consider the Stripe API documentation. It is a highly renowned platform for
+excellent documentation and existing client libraries which was developed by
+Stripe and their team. It is an invaluable asset and very powerful platform to
+Stripe. This project can democratize that power. This project can inexpensively
+provide excellent documentation and client libraries for web API creators.
+
+# Who is the target audience? How would they benefit from this project?
+
+Any creator of a web API which exposes a supported schema/hypermedia format.
```
Comment (tomchristie) [inline posts/okfn.md]: Perhaps "creator or user"

**Pull request closed without merging**